docs/extending/editor_api.md

```diff
@@ -0,0 +1,168 @@
+    this.renderFurniture();
+  }
+
+  renderFurniture() {
+    const prefix = this.element.closest('[id]').id;
+    const buttonId = `${prefix}-generate`;
+    const button = /* html */ `
+      <button
+        id="${buttonId}"
+        type="button"
+        data-summarize-target="suggest"
+        data-action="summarize#generate"
+        class="button"
+      >
+        Generate suggestions
+      </button>
+    `;
+    this.element.insertAdjacentHTML('beforeend', button);
+
+    this.outputArea = document.createElement('div');
+    this.element.append(this.outputArea);
+  }
+
+  renderSuggestion(suggestion) {
+    const template = document.createElement('template');
+    template.innerHTML = /* html */ `
+      <div>
+        <output for="${this.suggestTarget.id}">${suggestion}</output>
+        <button class="button button-small" type="button" data-action="summarize#useSuggestion">Use</button>
+      </div>
+    `;
+    this.outputArea.append(template.content.firstElementChild);
+  }
+
+  useSuggestion(event) {
+    this.input.value = event.target.previousElementSibling.textContent;
+  }
+
+  async summarize(text) {
+    const summarizer = await this.summarizer;
+    return summarizer.summarize(text);
+  }
+
+  async getPageContent() {
+    const previewController = window.wagtail.app.queryController('w-preview');
+    const { innerText, lang } = await previewController.extractContent();
+    this.contentLanguage = lang;
+    return innerText;
+  }
+
+  async generate() {
+    this.outputArea.innerHTML = '';
+    this.suggestTarget.textContent = 'Generating…';
+    this.suggestTarget.disabled = true;
+
+    const text = await this.getPageContent();
+    await Promise.allSettled(
+      [...Array(3).keys()].map(() =>
+        this.summarize(text)
+          .then((output) => this.renderSuggestion(output))
+          .catch((error) => {
+            console.error('Error generating suggestion:', error);
+          }),
+      ),
+    );
+
+    this.suggestTarget.disabled = false;
+    this.suggestTarget.textContent = 'Generate suggestions';
+    this.clearTarget.hidden = false; // Show the clear button
+  }
+}
+
+window.wagtail.app.register('summarize', SummarizeController);
+```
+
+The JavaScript file can be loaded to the editor using the `insert_editor_js` hook and attached to the `Page`'s `FieldPanel` for the `search_description` field:
+
+```python
+# myapp/wagtail_hooks.py
+from django.templatetags.static import static
+from wagtail import hooks
+from wagtail.admin.panels import FieldPanel
+from wagtail.models import Page
+
+
+@hooks.register("insert_editor_js")
+def editor_js():
+    js_files = ["js/summarize.js"]
+    return format_html_join(
+        "\n",
+        '<script src="{}"></script>',
+        ((static(filename),) for filename in js_files),
+    )
+
+# Replace the default `FieldPanel` for `search_description`
+# with a custom one that uses the `summarize` controller.
+Page.promote_panels[0].args[0][-1] = FieldPanel(
```

Comment:
A bit awkward after #12750 now that we have `PanelPlaceholder`s in the default panels rather than actual `FieldPanel`s.